File: app/src/main/java/soup/movie/ui/detail/DetailPresenter.java
package soup.movie.ui.detail;

import android.support.annotation.NonNull;
import android.util.Pair;

import java.util.Collections;
import java.util.List;

import javax.inject.Inject;

import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import soup.movie.data.MoobRepository;
import soup.movie.data.model.Movie;
import soup.movie.data.model.Theater;
import soup.movie.data.model.TimeTable;
import soup.movie.data.request.TimeTableRequest;
import soup.movie.data.response.TimeTableResponse;
import soup.movie.data.model.Trailer;
import soup.movie.di.scope.ActivityScoped;
import soup.movie.settings.TheaterSetting;
import soup.movie.ui.BasePresenter;
import timber.log.Timber;

@ActivityScoped
public class DetailPresenter extends BasePresenter<DetailContract.View>
        implements DetailContract.Presenter {

    private final MoobRepository moobRepository;
    private final TheaterSetting theaterSetting;

    @Inject
    DetailPresenter(MoobRepository moobRepository,
                    TheaterSetting theaterSetting) {
        this.moobRepository = moobRepository;
        this.theaterSetting = theaterSetting;
    }

    @Override
    public void requestData(@NonNull Movie movie) {
        List<Theater> theaters = theaterSetting.getFavoriteTheaters();
        //view.render(new DetailViewState.LoadingState(!theaters.isEmpty()));
        if (theaters.isEmpty()) {
            register(getTrailerListObservable(movie)
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(
                            trailers -> view.render(new DetailViewState.DoneState(
                                    new TimeTable(Collections.emptyList()), trailers)),
                            Timber::e));
        } else {
            register(Single.zip(
                    getTimeTableObservable(theaters.get(0).getCode(), movie.getId()),
                    getTrailerListObservable(movie),
                    Pair::create)
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(
                            pair -> view.render(new DetailViewState.DoneState(pair.first, pair.second)),
                            Timber::e));
        }
    }

    @Override
    public void requestData(@NonNull String code, @NonNull Movie movie) {
        register(Single.zip(
                getTimeTableObservable(code, movie.getId()),
                getTrailerListObservable(movie),
                Pair::create)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        pair -> view.render(new DetailViewState.DoneState(pair.first, pair.second)),
                        Timber::e));
    }

    private Single<TimeTable> getTimeTableObservable(@NonNull String theaterId, @NonNull String movieId) {
        return moobRepository
                .getTimeTableList(new TimeTableRequest(theaterId, movieId))
                .onErrorReturn(throwable -> new TimeTableResponse(new TimeTable(Collections.emptyList())))
                .map(TimeTableResponse::getTimeTable);
    }

    private Single<List<Trailer>> getTrailerListObservable(@NonNull Movie movie) {
        return Single.just(movie.getTrailers())
                .onErrorReturn(throwable -> Collections.emptyList());
    }
}

File: app/src/main/java/soup/movie/ui/detail/DetailPresenter.kt
package soup.movie.ui.detail

import android.annotation.SuppressLint
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.internal.disposables.DisposableContainer
import io.reactivex.schedulers.Schedulers
import soup.movie.R
import soup.movie.data.MovieSelectManager
import soup.movie.data.model.Movie
import soup.movie.data.model.Theater.Companion.TYPE_CGV
import soup.movie.data.model.Theater.Companion.TYPE_LOTTE
import soup.movie.data.model.Theater.Companion.TYPE_MEGABOX
import soup.movie.data.model.Theater.Companion.TYPE_NONE
import soup.movie.settings.impl.UsePaletteThemeSetting
import soup.movie.ui.BasePresenter
import soup.movie.ui.detail.DetailContract.Presenter
import soup.movie.ui.detail.DetailContract.View
import soup.movie.ui.detail.DetailViewState.DoneState
import soup.movie.ui.detail.DetailViewState.ListItem
import soup.movie.util.ImageUriProvider
import java.util.concurrent.TimeUnit

class DetailPresenter(private var usePaletteThemeSetting: UsePaletteThemeSetting,
                      private var imageUriProvider: ImageUriProvider) :
        BasePresenter<View>(), Presenter {

    override fun initObservable(disposable: DisposableContainer) {
        super.initObservable(disposable)
        disposable.add(MovieSelectManager
                .asObservable()
                .subscribeOn(Schedulers.io())
                .delay(500, TimeUnit.MILLISECONDS)
                .map { DoneState(it.toItems()) }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { view?.render(it) }
        )
    }

    @SuppressLint("CheckResult")
    override fun requestShareImage(url: String) {
        imageUriProvider(url)
                .subscribeOn(Schedulers.io())
                .subscribe { view?.doShareImage(it, getImageMimeType(url)) }
    }

    private fun getImageMimeType(fileName: String): String = when {
        fileName.endsWith(".png") -> "image/png"
        fileName.endsWith(".gif") -> "image/gif"
        else -> "image/jpeg"
    }

    override fun usePaletteTheme(): Boolean = usePaletteThemeSetting.get()

    private fun Movie.toItems(): List<ListItem> {
        val list = mutableListOf(
                ListItem(TYPE_CGV, R.layout.item_detail_cgv, this),
                ListItem(TYPE_LOTTE, R.layout.item_detail_lotte, this),
                ListItem(TYPE_MEGABOX, R.layout.item_detail_megabox, this))
        if (hasNaverInfo()) {
            list.add(ListItem(TYPE_NONE, R.layout.item_detail_naver, this))
        }
        list.add(ListItem(TYPE_NONE, R.layout.item_detail_trailers, this, trailers.orEmpty()))
        return list
    }
}
